File: bottomsheetmenu/src/main/java/com/krossovochkin/bottomsheetmenu/BottomSheetMenu.java
/*
 * Copyright (C) 2017 Vasya Drobushkov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.krossovochkin.bottomsheetmenu;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.BottomSheetDialog;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.lang.reflect.Constructor;

public class BottomSheetMenu extends BottomSheetDialog {

    private final BottomSheetMenuListener mBottomSheetMenuListener;
    private final int mIconSize;
    @Nullable
    private CharSequence mTitle;

    public static class Builder {

        private final Context mContext;
        private final BottomSheetMenuListener mBottomSheetMenuListener;
        private CharSequence mTitle;

        public Builder(@NonNull Context context, BottomSheetMenuListener listener) {
            mContext = context;
            mBottomSheetMenuListener = listener;
        }

        public BottomSheetMenu.Builder setTitle(@StringRes int titleId) {
            mTitle = mContext.getString(titleId);
            return this;
        }

        public BottomSheetMenu.Builder setTitle(CharSequence title) {
            mTitle = title;
            return this;
        }

        private BottomSheetMenu create() {
            final BottomSheetMenu menu = new BottomSheetMenu(mContext, mBottomSheetMenuListener);
            menu.setTitle(mTitle);
            menu.build();
            return menu;
        }

        @SuppressWarnings("UnusedReturnValue")
        public BottomSheetMenu show() {
            final BottomSheetMenu menu = create();
            menu.show();
            return menu;
        }
    }

    @SuppressLint("InflateParams")
    private BottomSheetMenu(@NonNull Context context, BottomSheetMenuListener bottomSheetMenuListener) {
        super(context);
        mIconSize = context.getResources().getDimensionPixelSize(R.dimen.bottom_sheet_menu_item_icon_size);
        mBottomSheetMenuListener = bottomSheetMenuListener;

        setOnShowListener(new DialogInterface.OnShowListener() {
            @Override
            public void onShow(DialogInterface dialog) {
                BottomSheetDialog d = (BottomSheetDialog) dialog;
                FrameLayout bottomSheet = (FrameLayout) d.findViewById(android.support.design.R.id.design_bottom_sheet);

                if (bottomSheet != null) {
                    BottomSheetBehavior.from(bottomSheet).setState(BottomSheetBehavior.STATE_EXPANDED);
                }
            }
        });
    }

    @Override
    public void setTitle(CharSequence title) {
        // no need to call super.setTitle, as title is blocked in BottomSheetDialog
        mTitle = title;
    }

    private void build() {
        final LayoutInflater layoutInflater = LayoutInflater.from(getContext());
        final Menu menu = newMenuInstance(getContext());

        if (menu == null) {
            return;
        }

        final MenuInflater menuInflater = new MenuInflater(getContext());
        mBottomSheetMenuListener.onCreateBottomSheetMenu(menuInflater, menu);

        final LinearLayout rootView = (LinearLayout) View.inflate(getContext(), R.layout.view_bottom_sheet_menu, null);
        rootView.addView(createHeaderView(layoutInflater, rootView, mTitle));
        for (int i = 0; i < menu.size(); i++) {
            rootView.addView(createMenuItemView(layoutInflater, rootView, menu.getItem(i)));
        }
        setContentView(rootView);
    }

    private View createHeaderView(final LayoutInflater inflater, ViewGroup rootView, @Nullable CharSequence title) {
        if (TextUtils.isEmpty(title)) {
            return inflater.inflate(R.layout.view_bottom_sheet_header_placeholder, rootView, false);
        } else {
            final TextView titleView = (TextView) inflater.inflate(R.layout.view_bottom_sheet_header, rootView, false);
            titleView.setText(mTitle);
            return titleView;
        }
    }

    private View createMenuItemView(final LayoutInflater inflater, ViewGroup rootView, final MenuItem item) {
        final TextView menuItemView = (TextView) inflater.inflate(R.layout.view_bottom_sheet_menu_item, rootView, false);

        menuItemView.setText(item.getTitle());

        final Drawable drawable = item.getIcon().mutate();
        drawable.setBounds(0, 0, mIconSize, mIconSize);
        menuItemView.setCompoundDrawables(drawable, null, null, null);

        menuItemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mBottomSheetMenuListener.onBottomSheetMenuItemSelected(item);
                dismiss();
            }
        });
        return menuItemView;
    }

    public interface BottomSheetMenuListener {
        void onCreateBottomSheetMenu(MenuInflater inflater, Menu menu);

        void onBottomSheetMenuItemSelected(MenuItem item);
    }

    @SuppressLint("PrivateApi")
    private Menu newMenuInstance(Context context) {
        try {
            Class<?> menuBuilderClass = Class.forName("com.android.internal.view.menu.MenuBuilder");
            Constructor<?> constructor = menuBuilderClass.getDeclaredConstructor(Context.class);
            return (Menu) constructor.newInstance(context);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: bottomsheetmenukt/src/main/java/com/krossovochkin/bottomsheetmenukt/BottomSheetMenu.kt
/*
 * Copyright (C) 2017 Vasya Drobushkov
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.krossovochkin.bottomsheetmenukt

import android.annotation.SuppressLint
import android.content.Context
import android.support.annotation.ColorInt
import android.support.annotation.StringRes
import android.support.design.widget.BottomSheetBehavior
import android.support.design.widget.BottomSheetDialog
import android.support.v4.graphics.drawable.DrawableCompat
import android.text.SpannableString
import android.text.Spanned
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import android.view.*
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import kotlin.properties.Delegates

fun MenuItem.setTextColor(@ColorInt textColor: Int) {
    val s = SpannableString(title)
    s.setSpan(ForegroundColorSpan(textColor), 0, s.length, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
    title = s
}

fun MenuItem.setIconTint(@ColorInt tintColor: Int) {
    val drawable = icon.mutate()
    DrawableCompat.setTint(drawable, tintColor)
    icon = drawable
}

class BottomSheetMenu(
        context: Context,
        private val bottomSheetMenuListener: BottomSheetMenuListener)
    : BottomSheetDialog(context) {

    private var iconSize : Int by Delegates.notNull()
    private var title : CharSequence? = null

    init {
        iconSize = context.resources.getDimensionPixelSize(R.dimen.bottom_sheet_menu_item_icon_size)

        setOnShowListener({ dialog ->
            val d = dialog as BottomSheetDialog
            val bottomSheet = d.findViewById(android.support.design.R.id.design_bottom_sheet) as FrameLayout
            BottomSheetBehavior.from(bottomSheet).setState(BottomSheetBehavior.STATE_EXPANDED)
        })
    }

    class Builder(private val context: Context, private val bottomSheetMenuListener: BottomSheetMenuListener) {

        private var title : CharSequence? = null

        fun setTitle(@StringRes titleId : Int) : Builder {
            title = context.getString(titleId)
            return this
        }

        fun setTitle(title : CharSequence?) : Builder {
            this.title = title
            return this
        }

        private fun create(): BottomSheetMenu {
            val menu = BottomSheetMenu(context, bottomSheetMenuListener)
            menu.setTitle(title)
            menu.build()
            return menu
        }

        fun show(): BottomSheetMenu {
            val menu = create()
            menu.show()
            return menu
        }
    }

    override fun setTitle(title: CharSequence?) {
        // no need to call super.setTitle, as title is blocked in BottomSheetDialog
        this.title = title
    }

    private fun build() {
        val layoutInflater = LayoutInflater.from(context)
        val menu = newMenuInstance(context) ?: return

        val menuInflater = MenuInflater(context)
        bottomSheetMenuListener.onCreateBottomSheetMenu(menuInflater, menu)

        val rootView = View.inflate(context, R.layout.view_bottom_sheet_menu, null) as LinearLayout
        rootView.addView(createHeaderView(layoutInflater, rootView, title))
        for (i in 0 until menu.size()) {
            rootView.addView(createMenuItemView(layoutInflater, rootView, menu.getItem(i)))
        }
        setContentView(rootView)
    }

    private fun createHeaderView(inflater: LayoutInflater, rootView: ViewGroup, title: CharSequence?): View {
        if (TextUtils.isEmpty(title)) {
            return inflater.inflate(R.layout.view_bottom_sheet_header_placeholder, rootView, false)
        } else {
            val titleView = inflater.inflate(R.layout.view_bottom_sheet_header, rootView, false) as TextView
            titleView.text = title
            return titleView
        }
    }

    private fun createMenuItemView(inflater: LayoutInflater, rootView: ViewGroup, item: MenuItem): View {
        val menuItemView = inflater.inflate(R.layout.view_bottom_sheet_menu_item, rootView, false) as TextView

        menuItemView.text = item.title

        val drawable = item.icon.mutate()
        drawable.setBounds(0, 0, iconSize, iconSize)
        menuItemView.setCompoundDrawables(drawable, null, null, null)

        menuItemView.setOnClickListener {
            bottomSheetMenuListener.onBottomSheetMenuItemSelected(item)
            dismiss()
        }
        return menuItemView
    }

    interface BottomSheetMenuListener {
        fun onCreateBottomSheetMenu(inflater: MenuInflater, menu: Menu)

        fun onBottomSheetMenuItemSelected(item: MenuItem)
    }

    @SuppressLint("PrivateApi")
    private fun newMenuInstance(context: Context): Menu? {
        try {
            val menuBuilderClass = Class.forName("com.android.internal.view.menu.MenuBuilder")
            val constructor = menuBuilderClass.getDeclaredConstructor(Context::class.java)
            return constructor.newInstance(context) as Menu
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return null
    }
}